File: desktop/src/domain/imports.ts
import type { MessageRole } from "./common";
import type { ContentBlock } from "./content";
import type { EvidenceRef } from "./evidence";
import type {
  GeneratorRef,
  KnowledgeEntityKind,
  KnowledgeScope,
} from "./knowledge";

export type ImportPlatform = "chatGpt" | "claude" | "codex" | "generic";
export type ImportRevisionStatus = "parsing" | "parsed" | "partiallyParsed" | "failed";
export type RecoveryStatus = "recovered" | "partial" | "unavailable";
export type ImportFormat = "markdown" | "jsonLines" | "text" | "json";
export type ImportIngress = "filePicker" | "dragAndDrop" | "paste";
export type ImportAnalysisPolicy = "disabled";

export type GenericImportDescriptor = {
  contractVersion: string;
  importSourceId: string;
  format: ImportFormat;
  ingress: ImportIngress;
  mediaType: string | null;
  encoding: string;
  byteLength: number;
  contentHash: string;
  immutableStorageRef: string;
};

export type RawTrackEntry = {
  id: string;
  importSourceId: string;
  importRevisionId: string;
  importedMessageId: string;
  sourceLocator: string;
  contentHash: string;
  ordinal: number;
};

export type ImportGraphProjection = {
  importSourceId: string;
  importRevisionId: string;
  conversationId: string;
  entryNodeId: string;
  rawTrackEntryIds: string[];
  analysisPolicy: ImportAnalysisPolicy;
};

export type ImportSource = {
  id: string;
  conversationId: string;
  platform: ImportPlatform;
  originalFileName: string | null;
  contentHash: string;
  storageRef: string;
  createdAt: string;
};

export type ImportRevision = {
  id: string;
  importSourceId: string;
  adapterId: string;
  adapterVersion: string;
  status: ImportRevisionStatus;
  createdAt: string;
};

export type FieldRecovery = {
  field: string;
  status: RecoveryStatus;
  detail: string | null;
};

export type ImportIssue = {
  code: string;
  message: string;
  sourceLocator: string | null;
  recoverable: boolean;
};

export type ParseReport = {
  importRevisionId: string;
  conversationCount: number;
  messageCount: number;
  attachmentCount: number;
  toolRecordCount: number;
  fieldRecovery: FieldRecovery[];
  warnings: ImportIssue[];
  errors: ImportIssue[];
};

export type ImportedMessage = {
  id: string;
  importRevisionId: string;
  role: MessageRole;
  contentBlocks: ContentBlock[];
  occurredAt: string | null;
  sourceLocator: string;
  parentImportedMessageId: string | null;
  platformExtension: unknown;
};

export type ImportBundleQueryProjection = {
  source: ImportSource;
  revision: ImportRevision;
  messages: ImportedMessage[];
  report: ParseReport;
};

export type DerivedContinuation = {
  id: string;
  importSourceId: string;
  importRevisionId: string;
  revision: number;
  analysisMode: "quick" | "detailed";
  generatorId: string;
  generatorVersion: string;
  status: "active" | "superseded" | "invalidated" | "deleted";
  claims: Array<{
    id: string;
    kind: string;
    value: string;
    evidence: EvidenceRef[];
    userConfirmed: boolean;
  }>;
  createdAt: string;
};

export type ImportKnowledgeProposalRequestInput = {
  requestId: string;
  importSourceId: string;
  importRevisionId: string;
  expectedSourceContentHash: string;
  selectedMessageIds: string[];
  targetScope: KnowledgeScope;
  requestedAt: string;
};

export type ImportKnowledgeEntityProposal = {
  contractVersion: "mindscape.import-knowledge-proposal.v1";
  proposalId: string;
  requestId: string;
  importSourceId: string;
  importRevisionId: string;
  conversationId: string;
  suggestedKind: KnowledgeEntityKind;
  suggestedName: string;
  suggestedAliases: string[];
  targetScope: KnowledgeScope;
  evidence: EvidenceRef[];
  generator: GeneratorRef;
  proposalRevision: number;
  proposedAt: string;
};

export type ImportKnowledgeProposalBatchProjection = {
  contractVersion: "mindscape.import-knowledge-proposal.v1";
  requestId: string;
  importSourceId: string;
  importRevisionId: string;
  conversationId: string;
  sourceContentHash: string;
  targetScope: KnowledgeScope;
  generationRunId: string;
  generator: GeneratorRef;
  proposals: ImportKnowledgeEntityProposal[];
  batchRevision: number;
  requestedAt: string;
  generatedAt: string;
};

export type ImportKnowledgeProposalReviewChoice =
  | {
      action: "confirm";
      kind: KnowledgeEntityKind;
      name: string;
      aliases: string[];
    }
  | { action: "reject"; reason: string | null };

export type ImportKnowledgeProposalReviewCommandInput = {
  decisionId: string;
  requestId: string;
  proposalId: string;
  expectedProposalRevision: number;
  choice: ImportKnowledgeProposalReviewChoice;
  decidedAt: string;
};

export type ImportKnowledgeProposalReviewProjection = {
  contractVersion: "mindscape.import-knowledge-proposal.v1";
  decisionId: string;
  requestId: string;
  proposalId: string;
  proposalRevision: number;
  action: "confirm" | "reject";
  entityId: string | null;
  entityStatus: "candidate" | "confirmed" | null;
  decidedBy: GeneratorRef;
  decidedAt: string;
};

File: desktop/src-tauri/src/domain/contracts/imports.rs
use serde::{Deserialize, Serialize};

use crate::domain::{ContentBlock, MessageRole};

use super::EvidenceRef;

pub const IMPORT_CONTRACT_VERSION: &str = "mindscape.import.v1";
pub const GENERIC_IMPORT_CONTRACT_VERSION: &str = "mindscape.generic-import.v1";

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ImportFormat {
    Markdown,
    JsonLines,
    Text,
    Json,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ImportIngress {
    FilePicker,
    DragAndDrop,
    Paste,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ImportAnalysisPolicy {
    Disabled,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GenericImportDescriptor {
    pub contract_version: String,
    pub import_source_id: String,
    pub format: ImportFormat,
    pub ingress: ImportIngress,
    pub media_type: Option<String>,
    pub encoding: String,
    pub byte_length: u64,
    pub content_hash: String,
    pub immutable_storage_ref: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RawTrackEntry {
    pub id: String,
    pub import_source_id: String,
    pub import_revision_id: String,
    pub imported_message_id: String,
    pub source_locator: String,
    pub content_hash: String,
    pub ordinal: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportGraphProjection {
    pub import_source_id: String,
    pub import_revision_id: String,
    pub conversation_id: String,
    pub entry_node_id: String,
    pub raw_track_entry_ids: Vec<String>,
    pub analysis_policy: ImportAnalysisPolicy,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ImportPlatform {
    ChatGpt,
    Claude,
    Codex,
    Generic,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportSource {
    pub id: String,
    pub conversation_id: String,
    pub platform: ImportPlatform,
    pub original_file_name: Option<String>,
    pub content_hash: String,
    pub storage_ref: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ImportRevisionStatus {
    Parsing,
    Parsed,
    PartiallyParsed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportRevision {
    pub id: String,
    pub import_source_id: String,
    pub adapter_id: String,
    pub adapter_version: String,
    pub status: ImportRevisionStatus,
    pub created_at: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum RecoveryStatus {
    Recovered,
    Partial,
    Unavailable,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FieldRecovery {
    pub field: String,
    pub status: RecoveryStatus,
    pub detail: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportIssue {
    pub code: String,
    pub message: String,
    pub source_locator: Option<String>,
    pub recoverable: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ParseReport {
    pub import_revision_id: String,
    pub conversation_count: u64,
    pub message_count: u64,
    pub attachment_count: u64,
    pub tool_record_count: u64,
    pub field_recovery: Vec<FieldRecovery>,
    pub warnings: Vec<ImportIssue>,
    pub errors: Vec<ImportIssue>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportedMessage {
    pub id: String,
    pub import_revision_id: String,
    pub role: MessageRole,
    pub content_blocks: Vec<ContentBlock>,
    pub occurred_at: Option<String>,
    pub source_locator: String,
    pub parent_imported_message_id: Option<String>,
    pub platform_extension: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportBundleQueryProjection {
    pub source: ImportSource,
    pub revision: ImportRevision,
    pub messages: Vec<ImportedMessage>,
    pub report: ParseReport,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum AnalysisMode {
    Quick,
    Detailed,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ContinuationStatus {
    Active,
    Superseded,
    Invalidated,
    Deleted,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContinuationClaim {
    pub id: String,
    pub kind: String,
    pub value: String,
    pub evidence: Vec<EvidenceRef>,
    pub user_confirmed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DerivedContinuation {
    pub id: String,
    pub import_source_id: String,
    pub import_revision_id: String,
    pub revision: u64,
    pub analysis_mode: AnalysisMode,
    pub generator_id: String,
    pub generator_version: String,
    pub status: ContinuationStatus,
    pub claims: Vec<ContinuationClaim>,
    pub created_at: String,
}
